### PR TITLE
Enhance csv version of output

### LINKED ISSUE #1026: Enhance csv version of output
# 💡 Summary #
Enhance the csv output to include details such as the requirement text and result values. Basically, make it a csv version of the reports.

## Motivation and context ##
While ScubaGear does output a csv version of the output, this functionality has not been updated in some time and does not include several enhancements that have been made. For example, it does not include the requirement text (#1015) or the "Result" value.

This would allow people to access their results in a machine-readable format or port their results to Excel spreadsheets.

## Implementation notes ##
PR #970 created a json version of the output. The easiest way to implement the csv enhancement would probably be to convert the ScubaResults.json "Results" key into a csv.

## Acceptance criteria ##
- [ ] A csv version of the output is available

### PR DESCRIPTION
**Pull request by adhilto**
#  #
  

## 🗣 Description ##
Remove the TestResults.csv file, which was just a csv version of the Rego output and replace it with a new csv which is structured the same way as the Results are in the html files.

## 💭 Motivation and context ##
Closes #1026

## 🧪 Testing ##
- I tested manually and reviewed output
- I added new PowerShell unit tests

## 📷 Screenshots (if appropriate) ##
The old TestResults.csv, which looked like this:
image

Is replaced with the new ScubaResults.csv file looks like this:
image

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [x] PR passed smoke test check.
- [x] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [x] Resolved all merge conflicts on branch
- [x] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Comment by james-garriss** [comment]
Hey, @adhilto , did you look at the documentation to see if it needs to be updated IAW your coding changes?

**Comment by james-garriss** [comment]
@adhilto , I noticed that you created Pester tests files for converting to CSV (ConvertTo-ResultsCsv.Tests.ps1) and for testing plain text formatting (Format-PlainText.Tests.ps1).  Breaking out these tests seems like a great idea.  It makes me wonder where the corresponding test files are for converting to other file formats and formatting them.  I don't see any other files similar to these 2 and I intuit they should exist.  If they should exist, but if they are out of scope for your issue, please create one or more other issues to ensure that we eventually get to feature testing parity.

**Comment by adhilto** [comment]
> @adhilto , I noticed that you created Pester tests files for converting to CSV (ConvertTo-ResultsCsv.Tests.ps1) and for testing plain text formatting (Format-PlainText.Tests.ps1). Breaking out these tests seems like a great idea. It makes me wonder where the corresponding test files are for converting to other file formats and formatting them. I don't see any other files similar to these 2 and I intuit they should exist. If they should exist, but if they are out of scope for your issue, please create one or more other issues to ensure that we eventually get to feature testing parity.

We do have test cases for the json (`Merge-JsonOutput.tests.ps1`). The other output format we have is HTML, which we do have tests for, but you have to look inside the "Create-Report" folder for those. So I think we're good on this front, but let me know if I'm missing anything.

**Comment by adhilto** [comment]
> Hey, @adhilto , did you look at the documentation to see if it needs to be updated IAW your coding changes?

Turns out there is, thanks for the reminder. I'll get that updated soon.

**Comment by adhilto** [comment]
> > Hey, @adhilto , did you look at the documentation to see if it needs to be updated IAW your coding changes?
> 
> Turns out there is, thanks for the reminder. I'll get that updated soon.

Done

**Comment by james-garriss** [comment]
In the docs I see this parameter: `OutJsonFileName`.  Should I also see a similar one for CSV?

**Comment by adhilto** [comment]
> In the docs I see this parameter: `OutJsonFileName`. Should I also see a similar one for CSV?

...yes. Let me try again.

**Comment by adhilto** [comment]
@james-garriss @buidav Ready for review again

**Comment by james-garriss** [comment]
TestResults.json needs to be added back the Reports page, as it's a report.

https://github.com/cisagov/ScubaGear/blob/1026-enhance-csv-version-of-output/docs/execution/reports.md

**Comment by james-garriss** [comment]
> Also, we should highly encourage the `-MergeJSON` flag then in the README

I still feel like we are missing the right parameter.  If someone wants a CSV output, then `-MergeJSON` is counterintuitive in the extreme.  Something like a simple `-CSV` flag seems better.

**Comment by james-garriss** [comment]
ScubaResults.json is now missing from the output, and there's the some weird repetitions in the output.  (@adhilto , see Slack for details).

**Comment by adhilto** [comment]
> > Also, we should highly encourage the `-MergeJSON` flag then in the README
> 
> I still feel like we are missing the right parameter. If someone wants a CSV output, then `-MergeJSON` is counterintuitive in the extreme. Something like a simple `-CSV` flag seems better.

Sorry, looks like I didn't explain the changes I made when I converted the PR back to a branch. The CSV output is **no longer** dependent on the `MergeJSON` flag. When I initially opened the PR it was, but after your feedback I removed that dependency and the CSV will _always_ be output.

**Review by james-garriss — APPROVED** [review]
Tested results, works great, ty.

**Comment by adhilto** [comment]
> Also, we should highly encourage the `-MergeJSON` flag then in the README. I would create a TODO issue for it as a placeholder until #1205 is complete.

Done https://github.com/cisagov/ScubaGear/issues/1286

**Comment by adhilto** [comment]
@nanda-katikaneni Ready for merge

**Comment by nanda-katikaneni** [comment]
Merging into the main as the smoke test error is a known issue (related to the current state of E5 tenant).
